src/runtime/CollectionWrappers/IterableWrapper.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Collections.Generic;
+using System.Collections;
+
+namespace Python.Runtime.CollectionWrappers
+{
+    internal class IterableWrapper<T> : IEnumerable<T>
+    {
+        protected readonly PyObject pyObject;
+
+        public IterableWrapper(PyObject pyObj)
+        {
+            if (pyObj == null)
+                throw new PythonException();
+            pyObject = pyObj;
+        }
+
+        private void propagateIterationException()
+        {
+            var err = Runtime.PyErr_Occurred();
+            if (err == null) return;
+
+            //remove StopIteration exceptions
+            if (0 != Runtime.PyErr_ExceptionMatches(Exceptions.StopIteration))
+            {
+                Runtime.PyErr_Clear();
+                return;
+            }
+
+            Runtime.CheckExceptionOccurred();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            PyObject iterObject = null;
+            using (Py.GIL())
```

Comment:
Actually, it is a job of the caller of this method to hold GIL for any call on `GetEnumerator` or `MoveNext`

Reply:
@lostmsu I defer to you but I am not convinced.  If I am passing an instance of a python list using this wrapper (let's say as an IEnumerable<T>) to some 3rd party library which is going to iterate over it at some point in the future, I would have to worry about when that library actually does the enumeration and enclose that with holding the GIL.

Personally I think it makes more sense to consider that the contract of IEnumerable<T> is "An object which can be iterated".  I think adding holding the GIL to the expectation of whoever is given an IEnumerable<T> that happens to be an IterableWrapper<T> isn't idiomatic C#.

Reply:
> it at some point in the future, I would have to worry about when that library actually does the enumeration and enclose that with holding the GIL.

I think you're right here.